add enum support for unions (fixes portion of #92)

aaronkub opened the pull request:
@marcosschroh This likely still needs a decent amount of work, but I figured it was best to open this now and get your thoughts.

**Overview**
* ~Rename current `EnumField` implementation to `SimpleEnumField`.~
    * Removed legacy `types.Enum` altogether
* `EnumField` now used for `enum.Enum` types.
* `_get_meta_class_attributes` method pulls any attributes from a `Meta` subclass in the enum (if it exists).
    * Totally open to a more elegant way of doing this.
* `_deserialize_complex_types` maps deserialized strings to the enum instance.
    * I believe that this implementation would fail in the case of a `Union[SomeEnum, str]` type. Not sure if this case needs to be supported.


**Questions**
* ~Do we want to remove `types.Enum` altogether? or leave it for now to avoid a breaking change?~
    * Removed

Open to all of your feedback on my current progress here!

marcosschroh commented:
Hi @aaronkub 

I will take a look asap. I think we should remove `SimpleEnumField`, otherwise It would be confusing having 2 ways of defining `enums`. We can alert users of the breaking change in the `CHANGELOG`, also we have a [migration guide](https://marcosschroh.github.io/dataclasses-avroschema/migration_guide/) where we can describe the situation.

aaronkub commented:
> Hi @aaronkub
> 
> I will take a look asap. I think we should remove `SimpleEnumField`, otherwise It would be confusing having 2 ways of defining `enums`. We can alert users of the breaking change in the `CHANGELOG`, also we have a [migration guide](https://marcosschroh.github.io/dataclasses-avroschema/migration_guide/) where we can describe the situation.

Sounds good! I'll get that `types.Enum` and `SimpleEnumField` removed.

marcosschroh commented on dataclasses_avroschema/schema_generator.py:
```diff
@@ -67,10 +68,44 @@ def get_fields(cls: typing.Any) -> typing.List[FieldType]:
+                enum_types[field_type.name] = field_type.type
+            elif isinstance(field_type, UnionField):
+                for sub_type in field_type.type.__args__:
+                    if issubclass(sub_type, enum.Enum):
```
Amazing job @aaronkub !! I see one issue in `deserialization`:

```python
import datetime
import uuid
import typing

from dataclasses_avroschema import AvroModel, types


class UnionSchema(AvroModel):
    "Some Unions"
    logical_union: typing.Union[datetime.datetime, datetime.date, uuid.uuid4]

instance = UnionSchema.fake()

# GOOD!!
In [67]: instance.serialize()
Out[67]: b'\x04H664f2e44-5d63-46ca-95a9-c4ae6d5d6d70'


In [70]: UnionSchema.deserialize(instance.serialize())
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-70-f795bc1d86fa> in <module>
----> 1 UnionSchema.deserialize(instance.serialize())

~/Projects/dataclasses-schema-generator/dataclasses_avroschema/schema_generator.py in deserialize(cls, data, serialization_type, create_instance, writer_schema)
    140         )
    141
--> 142         output = cls._deserialize_complex_types(payload)
    143
    144         if create_instance:

~/Projects/dataclasses-schema-generator/dataclasses_avroschema/schema_generator.py in _deserialize_complex_types(cls, payload)
     86     def _deserialize_complex_types(cls, payload: typing.Dict[str, typing.Any]) -> typing.Any:
     87         output = {}
---> 88         enum_type_map = cls._get_enum_type_map()
     89         for field, value in payload.items():
     90             if isinstance(value, dict):

~/Projects/dataclasses-schema-generator/dataclasses_avroschema/schema_generator.py in _get_enum_type_map(cls)
     77             elif isinstance(field_type, UnionField):
     78                 for sub_type in field_type.type.__args__:
---> 79                     if issubclass(sub_type, enum.Enum):
     80                         enum_types[field_type.name] = sub_type
     81             elif isinstance(field_type, RecordField):

TypeError: issubclass() arg 1 must be a class
```

aaronkub replied:
Thanks @marcosschroh! I will get this resolved ASAP.

aaronkub replied:
I've resolved this issue and added in your example as a unit test.

pawelrubin commented:
How to specify the default enum value? According to [Avro specification](https://avro.apache.org/docs/1.10.2/spec.html#Enums)
>default: A default value for this enumeration, used during resolution when the reader encounters a symbol from the writer that isn't defined in the reader's schema (optional). The value provided here must be a JSON string that's a member of the symbols array. See documentation on schema resolution for how this gets used.

aaronkub commented:
> How to specify the default enum value? According to [Avro specification](https://avro.apache.org/docs/1.10.2/spec.html#Enums)
> 
> > default: A default value for this enumeration, used during resolution when the reader encounters a symbol from the writer that isn't defined in the reader's schema (optional). The value provided here must be a JSON string that's a member of the symbols array. See documentation on schema resolution for how this gets used.

@pawelrubin Here is an example of defaults from the tests:

Python Class
```python
@pytest.fixture
def user_advance_with_defaults_dataclass(color_enum):
    @dataclasses.dataclass
    class UserAdvance(AvroModel):
        name: str
        age: int
        pets: typing.List[str] = dataclasses.field(default_factory=lambda: ["dog", "cat"])
        accounts: typing.Dict[str, int] = dataclasses.field(default_factory=lambda: {"key": 1})
        has_car: bool = False
        favorite_colors: color_enum = color_enum.BLUE
        country: str = "Argentina"
        address: str = None

        class Meta:
            schema_doc = False

    return UserAdvance
```
Avro Schema
```json
{
    "type": "record",
    "name": "UserAdvance",
    "fields": [
      {
        "name": "name",
        "type": "string"
      },
      {
        "name": "age",
        "type": "long"
      },
      {
        "name": "pets",
        "type": {
          "type": "array",
          "items": "string",
          "name": "pet"
        },
        "default": ["dog", "cat"]
      },
      {
        "name": "accounts",
        "type": {
          "type": "map",
          "values": "long",
          "name": "account"
        },
        "default": {
            "key": 1
        }
      },
      {
        "name": "has_car",
        "type": "boolean",
        "default": false
      },
      {
        "name": "favorite_colors",
        "type": {
          "type": "enum",
          "name": "favorite_color",
          "symbols": ["BLUE", "YELLOW", "GREEN"]
        },
        "default": "BLUE"
      },
      {
        "name": "country",
        "type": "string",
        "default": "Argentina"
      },
      {
        "name": "address",
        "type": ["null", "string"],
        "default": null
      }
    ]
}
``` 


This test asserts that the above model equals the schema.
```python
def test_schema_with_complex_types_and_defaults(
        user_advance_with_defaults_dataclass, 
        user_advance_with_defaults_avro_json):
    assert user_advance_with_defaults_dataclass.avro_schema() == json.dumps(user_advance_with_defaults_avro_json)
```

Does this answer your question? Let me know if I've misunderstood.

pawelrubin commented:
@aaronkub Well, not exactly. Let me rephrase my question. The enum type itself can have a `default` field. It's mentioned in the [**Avro specification**](https://avro.apache.org/docs/1.10.2/spec.html#Enums). The `default` value is used for schema resolution when the reader's schema does not contain a symbol from the writer's schema. In such case, the `default` value will be used.

Let me give an example, with **fastavro** as a reference.

```python
import io
import fastavro

writer_schema = {
    "type": "record",
    "name": "EnumTest",
    "fields": [
        {
            "name": "nullable_my_enum",
            "type": [
                "null",
                {
                    "type": "enum",
                    "name": "my_enum",
                    "symbols": ["FOO", "BAR"],
                },
            ],
            "default": None,
        }
    ],
}

reader_schema = {
    "type": "record",
    "name": "EnumTest",
    "fields": [
        {
            "name": "nullable_my_enum",
            "type": [
                "null",
                {
                    "type": "enum",
                    "name": "my_enum",
                    "symbols": ["FOO", "BUZZ"],
                    "default": "FOO",
                },
            ],
            "default": None,
        }
    ],
}

writer_record = {"nullable_my_enum": "BAR"}

output_bytes = io.BytesIO()

fastavro.schemaless_writer(output_bytes, writer_schema, writer_record)

input_bytes  = io.BytesIO(output_bytes.getvalue())

print(fastavro.schemaless_reader(input_bytes, writer_schema, reader_schema))

```

Note, that there are **two** `default`s in the reader's schema: one for the `nullable_my_enum` field and one as an attribute of `my_enum` type. More importantly, the reader's schema **does not** contain `"BAR"` symbol, therefore it is required to specify a default to successfully resolve the conflict between those two schemas.

This is also mentioned in the [**Avro specification**](https://avro.apache.org/docs/1.10.2/spec.html#Schema+Resolution) on schema resolution:
> if the writer's symbol is not present in the reader's enum and the reader has a default value, then that value is used, otherwise an error is signalled.

##

The output of the example code above is
```python
{'nullable_my_enum': 'FOO'}
```

And if there is no default, then an error will be signaled:
```python
Traceback (most recent call last):
  File "/Users/pawelrubin/vewd/czkafka/example.py", line 50, in <module>
    print(fastavro.schemaless_reader(input_bytes, writer_schema, reader_schema))
  File "fastavro/_read.pyx", line 1158, in fastavro._read.schemaless_reader
  File "fastavro/_read.pyx", line 1170, in fastavro._read.schemaless_reader
  File "fastavro/_read.pyx", line 820, in fastavro._read._read_data
  File "fastavro/_read.pyx", line 699, in fastavro._read.read_record
  File "fastavro/_read.pyx", line 812, in fastavro._read._read_data
  File "fastavro/_read.pyx", line 603, in fastavro._read.read_union
  File "fastavro/_read.pyx", line 794, in fastavro._read._read_data
  File "fastavro/_read.pyx", line 399, in fastavro._read.read_enum
fastavro._read_common.SchemaResolutionError: BAR not found in reader symbol list ['FOO', 'BUZZ']
```

aaronkub commented:
> @aaronkub Well, not exactly. Let me rephrase my question. The enum type itself can have a `default` field. It's mentioned in the [**Avro specification**](https://avro.apache.org/docs/1.10.2/spec.html#Enums). The `default` value is used for schema resolution when the reader's schema does not contain a symbol from the writer's schema. In such case, the `default` value will be used.
> 
> Let me give an example, with **fastavro** as a reference.
> 
> ```python
> import io
> import fastavro
> 
> writer_schema = {
>     "type": "record",
>     "name": "EnumTest",
>     "fields": [
>         {
>             "name": "nullable_my_enum",
>             "type": [
>                 "null",
>                 {
>                     "type": "enum",
>                     "name": "my_enum",
>                     "symbols": ["FOO", "BAR"],
>                 },
>             ],
>             "default": None,
>         }
>     ],
> }
> 
> reader_schema = {
>     "type": "record",
>     "name": "EnumTest",
>     "fields": [
>         {
>             "name": "nullable_my_enum",
>             "type": [
>                 "null",
>                 {
>                     "type": "enum",
>                     "name": "my_enum",
>                     "symbols": ["FOO", "BUZZ"],
>                     "default": "FOO",
>                 },
>             ],
>             "default": None,
>         }
>     ],
> }
> 
> writer_record = {"nullable_my_enum": "BAR"}
> 
> output_bytes = io.BytesIO()
> 
> fastavro.schemaless_writer(output_bytes, writer_schema, writer_record)
> 
> input_bytes  = io.BytesIO(output_bytes.getvalue())
> 
> print(fastavro.schemaless_reader(input_bytes, writer_schema, reader_schema))
> ```
> 
> Note, that there are **two** `default`s in the reader's schema: one for the `nullable_my_enum` field and one as an attribute of `my_enum` type. More importantly, the reader's schema **does not** contain `"BAR"` symbol, therefore it is required to specify a default to successfully resolve the conflict between those two schemas.
> 
> This is also mentioned in the [**Avro specification**](https://avro.apache.org/docs/1.10.2/spec.html#Schema+Resolution) on schema resolution:
> 
> > if the writer's symbol is not present in the reader's enum and the reader has a default value, then that value is used, otherwise an error is signalled.
> 
> The output of the example code above is
> 
> ```python
> {'nullable_my_enum': 'FOO'}
> ```
> 
> And if there is no default, then an error will be signaled:
> 
> ```python
> Traceback (most recent call last):
>   File "/Users/pawelrubin/vewd/czkafka/example.py", line 50, in <module>
>     print(fastavro.schemaless_reader(input_bytes, writer_schema, reader_schema))
>   File "fastavro/_read.pyx", line 1158, in fastavro._read.schemaless_reader
>   File "fastavro/_read.pyx", line 1170, in fastavro._read.schemaless_reader
>   File "fastavro/_read.pyx", line 820, in fastavro._read._read_data
>   File "fastavro/_read.pyx", line 699, in fastavro._read.read_record
>   File "fastavro/_read.pyx", line 812, in fastavro._read._read_data
>   File "fastavro/_read.pyx", line 603, in fastavro._read.read_union
>   File "fastavro/_read.pyx", line 794, in fastavro._read._read_data
>   File "fastavro/_read.pyx", line 399, in fastavro._read.read_enum
> fastavro._read_common.SchemaResolutionError: BAR not found in reader symbol list ['FOO', 'BUZZ']
> ```

@pawelrubin Thanks for the detailed reply! I believe I understand now. I wasn't aware of this Avro feature. It's not immediately clear to me what the best way to implement this is, but I'll play around with it. In the meantime, I'm open to any ideas you have.

aaronkub commented:
How about something like this, @pawelrubin? If that `get_default` method exists on the enum type we'll add it to the schema.

```python
def test_enum_field_with_type_level_default():
    class FavoriteColor(enum.Enum):
        BLUE = "Blue"
        GREEN = "Green"
        YELLOW = "Yellow"

        @classmethod
        def get_default(cls) -> 'FavoriteColor':
            return cls.GREEN

        class Meta:
            aliases = ["one", "two"]
            doc = "favorite colors"
            namespace = "some.name.space"

    enum_field = EnumField("field_name", FavoriteColor, FavoriteColor.BLUE, metadata={"key": "value"})

    assert enum_field.get_avro_type() == {
        "type": "enum",
        "name": "field_name",
        "symbols": ["Blue", "Green", "Yellow"],
        "aliases": ["one", "two"],
        "doc": "favorite colors",
        "namespace": "some.name.space",
        "default": "Green"
    }
```

pawelrubin commented:
How about a `Meta` attribute instead? I think it would be much cleaner and more intuitive.
```python
class FavoriteColor(enum.Enum):
    BLUE = "Blue"
    GREEN = "Green"
    YELLOW = "Yellow"

    class Meta:
        aliases = ["one", "two"]
        doc = "favorite colors"
        namespace = "some.name.space"
        default = "Yellow"
```

On has to use an explicit value instead of an `Enum` attribute, but that's not a big issue IMO.

aaronkub commented:
> How about a `Meta` attribute instead? I think it would be much cleaner and more intuitive.
> 
> ```python
> class FavoriteColor(enum.Enum):
>     BLUE = "Blue"
>     GREEN = "Green"
>     YELLOW = "Yellow"
> 
>     class Meta:
>         aliases = ["one", "two"]
>         doc = "favorite colors"
>         namespace = "some.name.space"
>         default = "Yellow"
> ```
> 
> On has to use an explicit value instead of an `Enum` attribute, but that's not a big issue IMO.

I agree that this is a better solution than what I have. I was stuck on the idea of using the enum instance, but I'll just add a verification step to make sure the supplied default is valid.

aaronkub commented:
@marcosschroh @pawelrubin Thanks for your comments! I'm pretty sure I've addressed the concerns you both had. Let me know if there's anything else you think should be changed.

marcosschroh commented:
Thanks @aaronkub for the `PR` and @pawelrubin  for the feedback. Looks good for me. Any other remarks @pawelrubin ?

pawelrubin commented on dataclasses_avroschema/fields.py:
```diff
@@ -460,37 +463,52 @@ def fake(self) -> bytes:
+        meta_dict = {}
+        if "Meta" in self.type.__members__:
+            meta = self.type.__members__["Meta"].value
+            meta_dict = {k: v for k, v in vars(meta).items() if k in self._VALID_META_FIELDS}
```
This is a clever solution to collect metadata for a field. However, a `dataclass` approach would be simpler in my opinion.

Consider something similar to the mechanism of collecting metadata in the `AvroModel`:
```python
CT = TypeVar("CT", bound="FieldMetadata")


@dataclass
class FieldMetadata:
    aliases: Optional[list[str]] = None
    default: Optional[str] = None
    doc: Optional[str] = None
    namespace: Optional[str] = None

    @classmethod
    def create(cls: type[CT], klass: type) -> CT:
        return cls(
            aliases=getattr(klass, "aliases", None),
            default=getattr(klass, "default", None),
            doc=getattr(klass, "doc", None),
            namespace=getattr(klass, "namespace", None),
        )
```

However, it would probably be worth refactoring other fields to use this approach. It might be out of the scope of this PR. What do you guys think?

marcosschroh replied:
Yes, I completely agree that the approach that you point is better, however I think we should  do it in another `PR`. We will need this at the moment of refactor the `Decimal` and `Fixed` types if I am not mistaken

pawelrubin commented on dataclasses_avroschema/fields.py:
```diff
@@ -460,37 +463,52 @@ def fake(self) -> bytes:
-            ), f"The default value should be on of {self.default.symbols}. Current is {default}"
-            return default
+                self.default.value in self.get_symbols()
+            ), f"The default value should be one of {self.get_symbols()}. Current is {self.default}"
```
Consider a priettier join.
```suggestion
            ), f"The default value should be one of {', '.join(self.get_symbols())}. Current is {self.default}"
```

pawelrubin commented on docs/complex_types.md:
````diff
@@ -17,34 +17,60 @@ The following list represent the avro complext types mapped to python types:
+```
+
+Note that in addition to a *field-level* default value, you can optionally supply a *type-level* default symbol for the 
+enum by supplying the `default` attribute under the `Meta` class.
````
I wonder if it would be worth mentioning the Avro specification on this. What do you think?

marcosschroh replied:
I think we should mention it.

aaronkub replied:
done!

pawelrubin commented on docs/redis_examples.md:
```diff
@@ -10,12 +11,17 @@ from walrus import Database  # A subclass of the redis-py Redis client.
     name: str
     age: int
-    favorite_colors: types.Enum = types.Enum(["BLUE", "YELLOW", "GREEN"], default="BLUE")
+    favorite_colors: FavoriteColor = FavoriteColor.BLUE
```
Make the name singular or add a `list` type.
```suggestion
    favorite_color: FavoriteColor = FavoriteColor.BLUE
```

pawelrubin commented on tests/schemas/test_faust_schema.py:
```diff
@@ -25,14 +26,14 @@ class Meta:
         age: int
         pets: typing.List[str]
         accounts: typing.Dict[str, int]
+        favorite_colors: color_enum
```
Here and other similar occurrences of `favorite_colors` as well, please 🙏🏻

pawelrubin commented on tests/schemas/test_faust_schema.py:
```diff
@@ -209,13 +208,21 @@ class Car(faust.Record, AvroModel):
         lake_trip: typing.Union[Bus, Car]
         river_trip: typing.Union[Bus, Car] = None
         mountain_trip: typing.Union[Bus, Car] = dataclasses.field(default_factory=lambda: {"engine_name": "honda"})
+        trip_distance: typing.Union[int, TripDistance] = None
```
I think this type annotation is incorrect. Since the default value is `None`, the type should be `Optional`.
```suggestion
        trip_distance: typing.Union[int, TripDistance, None] = None
```

aaronkub replied:
done!

pawelrubin commented on tests/schemas/test_schema.py:
```diff
@@ -107,3 +107,56 @@ class Aclass:
+    }
+
+
+def test_get_enum_type_map_with_sub_record(user_advance_dataclass_with_sub_record_and_enum, color_enum, user_type_enum):
```
Consider parametrizing those tests, as they are very similar.
```python
@pytest.mark.parametrize(
     "schema_cls",
     [user_advance_dataclass_with_enum, ...]. # Note: unfortunately, one cannot use fixtures here. Concrete classes have to be passed here.
)
def test_get_enum_type_map(schema_cls: AvroModel):
    assert schema_cls._get_enum_type_map() == {
        "favorite_colors": color_enum,
        "user_type": user_type_enum,
    }

aaronkub replied:
@pawelrubin It's not clear to me how to parametrize these tests without pulling the classes out of the fixtures. Is there a cleaner option you're aware of?

pawelrubin replied:
On second thought, it could make things more complicated. Too bad that **pylint** does not support using fixtures in `pytest.mark.parametrize`. However, you could try using [this](https://github.com/pytest-dev/pytest/issues/349#issuecomment-189370273) approach. It's a bit hacky though. Up to you.

aaronkub replied:
I think I'll opt to leave these tests as is then if we're all okay with that.

pawelrubin commented:
@aaronkub Looking good! 🔥 
I have just a few minor comments. Have a look, please.

pawelrubin commented:
I've tried out the following example:
```python
from typing import Optional
from dataclasses import dataclass
from enum import Enum

from dataclasses_avroschema import AvroModel


class MyEnum(str, Enum):
    A = "a"
    B = "b"

    class Meta:
        default = "a"


@dataclass
class Schema(AvroModel):
    enum_field: Optional[MyEnum] = None

print(Schema.avro_schema())
```

Unfortunately, it throws an exception:
```shell
<hidden>/dataclasses_avroschema/fields.py in _get_meta_class_attributes(self)
    470         if "Meta" in self.type.__members__:
    471             meta = self.type.__members__["Meta"].value
--> 472             meta_dict = {k: v for k, v in vars(meta).items() if k in self._VALID_META_FIELDS}
    473 
    474         default = meta_dict.get("default")

TypeError: vars() argument must have __dict__ attribute
```

I've changed the code to resolve this issue:
```python
def _get_meta_class_attributes(self) -> typing.Dict[str, typing.Any]:
    meta_dict = {}
    meta = getattr(self.type, "Meta", None)
    if meta is not None:
        meta_dict = {key: getattr(meta, key) for key in dir(meta) if key in self._VALID_META_FIELDS}
    
    ...
```

However, the resulting schema does not contain the default value for the enum. Additionally, the `"doc"` field is messy.
```python
>>> Schema.avro_schema()
'{"type": "record", "name": "Schema", "fields": [{"name": "enum_field", "type": ["null", {"type": "enum", "name": "MyEnum", "symbols": ["a", "b"]}], "default": null}], "doc": "Schema(enum_field: Union[__main__.MyEnum, NoneType] = None)"}'

aaronkub commented:
> I've tried out the following example:
> 
> ```python
> from typing import Optional
> from dataclasses import dataclass
> from enum import Enum
> 
> from dataclasses_avroschema import AvroModel
> 
> 
> class MyEnum(str, Enum):
>     A = "a"
>     B = "b"
> 
>     class Meta:
>         default = "a"
> 
> 
> @dataclass
> class Schema(AvroModel):
>     enum_field: Optional[MyEnum] = None
> 
> print(Schema.avro_schema())
> ```
> 
> Unfortunately, it throws an exception:
> 
> ```shell
> <hidden>/dataclasses_avroschema/fields.py in _get_meta_class_attributes(self)
>     470         if "Meta" in self.type.__members__:
>     471             meta = self.type.__members__["Meta"].value
> --> 472             meta_dict = {k: v for k, v in vars(meta).items() if k in self._VALID_META_FIELDS}
>     473 
>     474         default = meta_dict.get("default")
> 
> TypeError: vars() argument must have __dict__ attribute
> ```
> 
> I've changed the code to resolve this issue:
> 
> ```python
> def _get_meta_class_attributes(self) -> typing.Dict[str, typing.Any]:
>     meta_dict = {}
>     meta = getattr(self.type, "Meta", None)
>     if meta is not None:
>         meta_dict = {key: getattr(meta, key) for key in dir(meta) if key in self._VALID_META_FIELDS}
>     
>     ...
> ```
> 
> However, the resulting schema does not contain the default value for the enum. Additionally, the `"doc"` field is messy.
> 
> ```python
> >>> Schema.avro_schema()
> '{"type": "record", "name": "Schema", "fields": [{"name": "enum_field", "type": ["null", {"type": "enum", "name": "MyEnum", "symbols": ["a", "b"]}], "default": null}], "doc": "Schema(enum_field: Union[__main__.MyEnum, NoneType] = None)"}'
> ```

@pawelrubin Apologies for my delay in responding! This is an interesting use case. It seems that the inner `Meta` class gets converted to a string. It's not clear to me how to get the class back. Is multiple subclassing like this a common thing?

I may be missing a clear way to get the Meta attributes. Let me know if you're aware of how to do it. If this is something we want to support we may also need a different way to supply the field attributes.

marcosschroh commented:
Thanks @aaronkub  for you work? DO you want to update your changes? If you are too busy I can continue with the fix

marcosschroh commented:
closed by https://github.com/marcosschroh/dataclasses-avroschema/pull/153

Thanks for you work and effort @aaronkub !!
